crypto/cryptlib.c

```diff
@@ -272,6 +294,26 @@ void OPENSSL_showfatal(const char *fmta, ...)
     va_end(ap);
 
 # if defined(_WIN32_WINNT) && _WIN32_WINNT>=0x0333
+#  ifdef OPENSSL_SYS_WIN_CORE
+    /* ONECORE is always NONGUI and NT >= 0x0601 */
+    if (SUCCEEDED(TraceLoggingRegister(g_hProvider))) {
```

Comment:
its a good title for a book :) "Predict Unpredictable"
If I get it, that function is called when some critical error occur, not very common situation,
I just wonder what is the chance of overlapping here, and how it can impact on main purpose of this function. Calling TraceLoggingUnregister by some thread, indicates that the trace log has been already sent (its a first thread accessing it we assume it works at the start). There is slight possibility that other threads (if there one) will fail due to unexpected unregistration, also they must fail exactly in the same moment as the previous one.
First thing: It can be good idea to move debug output after Unregister, to shorten that time frame, as writing to output is always time consuming.

Reply:
As already said, reasoning "this is unlikely to happen/chance is slim" is *not* a legitimate excuse for not doing what's right. Suggestions like "let's make it shorter to minimize the window for failure" are *last* one to pose, not first. And only after one established that there is no way to avoid the unpredictable situation. Once again, can one place the structure on stack, i.e. not have in global?

Reply:
> not have it global

I assume that "register a provider" is kind of similar to opening a file. In sense the you get kind of a handle. And unpredictability would rather stem from the fact that in parallel attempt you would overwrite some structures from first attempt. This assumption might be wrong, but at least it's coherent with natural assumption that two threads should be able to log traces under same GUID independently of each other. Is it wrong assumption?

Reply:
Its weird tho, I simulated that situation on side:
Each of threads registered handler with same guid,then immediately sent event1
then sleeped for 15 seconds and then sent event2(diff msg) and immediately unregistered.
Second process started around 7 sec after the first one, to make sure process look like:
Process: A and B
A_Register
A_WriteEvent_1
(7 sec later)
B_Register
B_WriteEvent_1
(another 7 sec)
A_WriteEvent_2
A_Unregister
(another 7 sec)
B_WriteEvent_2 (It worked! didn't expected that)
B_Unregister
ETL file shows 4 events separated by around 7 sec delay in order A_1,B_1,A_2,B_2
It seems that unregistering from instance A hasn't affected event from B at all...

Reply:
> Its weird tho

I don't follow, what's weird? It worked as you intuitively expect a logging facility would work, entries appearing in order of submission. And I don't follow why wouldn't B_WriteEvent_2 work? When I wrote "unpredictability would rather stem from the fact that in parallel attempt you would overwrite some structures from first attempt" I meant specifically that you'd use same *global* variable to hold "kind of a handle" in same *process* context, e.g. referred by two threads. Note that it doesn't actually have to be two threads, you might as well be able to test with single thread, register(global), log(global), register(global), log(global), unregister(global), unregister(global). However, it makes lesser sense to test this, because documentation says "unpredictable" which doesn't actually preclude "expected" in some specific chain of events [say like just mentioned]. What's more important to test is register(variable1), log(valiable1), register(variable2), log(variable2), unregister(variable1), unregister(variable2), as well as some permutation of the sequence such as say register(variable1), registers(variable2), log(variable1), unregister(variable1), log(variable2), unregister(variable2).

Reply:
Just worried if registering handler with same name and guid by another process will have impact

Reply:
We can put some critical section here to make sure no thread will intercept procedure
but it needs to be declared and initialized in some common section before threads starts.
There is some functionality in crypto/threads_win.c.
Have idea If and how it can be used in here?

Reply:
> We can put some critical section here

Does it actually mean that structure can't be allocated at stack? Recall I said "some would even argue that it's inappropriate to even attempt to produce *a* message." Why is that? If you've got to this point as result of attack, you producing the message might be the part of their plan. In other words code shouldn't be more complex than minimally required. Yes, there is contradiction, because registering event source can be viewed as complex, yet it doesn't have to be more complex than it has to be.